# Exploring eBay Car Sales Data

## Table of Contents
---
## 1. **[Introduction_](#Introduction)**
## 2. **[Cleaning Column Names_](#Cleaning-Column-names)**
## 3. **[Initial Exploration and Cleaning_](#Initial-Exploration-and-Cleaning)**
## 4. **[Exploring the Odometer and Price Columns_](#Exploring-the-Odometer-and-Price-Columns)**
## 5. **[Exploring the date columns_](#Exploring-the-date-columns)**
## 6. **[Dealing with Incorrect Registration Year Data_](#Dealing-with-Incorrect-Registration-Year-Data)**
## 7. **[Exploring Price by Brand_](#Exploring-Price-by-Brand)**
## 8. **[Storing Aggregate Data in a DataFrame_](#Storing-Aggregate-Data-in-a-DataFrame)**


## Introduction

In this project, We are going to work with dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The dataset was originally uploaded on Kaggle by user [orgesleka](https://www.kaggle.com/orgesleka), but it's not available there anymore. You can access the dataset [here](https://data.world/data-society/used-cars-data).
The dataset we are going to use is a less cleaned version of the data, with 50,000 data points

The data dictionary provided with data is as follows:

* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which which year the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `kilometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which which year the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

Our aim in this project is to clean the data and analyze used car listings.

In [1]:
#Let's import all the libraries we are going to need in this project
import pandas as pd
import numpy as np

In [2]:
autos=pd.read_csv('autos.csv',encoding='Latin-1')
autos.info() # information of our dataframe
autos.head() # First Five rows of our dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

We can see the information about our dataset, As we mentioned in the introduction, We have 50000 rows in the dataset. There are 20 columns in the dataset and their data types are mentioned above. Some columns have null values, However most columns dont.

## Cleaning Column names
---
We will start our cleaning now, and the first thing we need are the column names. It is essential so to make sure the names of our columns dont have any problem and are easy to access. Let's check out our column names closely.

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We can see that the names of our columns are in `CamelCase` and not in `snake_case`, which is preferred in Python, which means we can't replace our spaces with underscores. So, We will reword some of column names based on the data dictionary to be more descriptive.

In [4]:
autos.columns=['date_crawled','name','seller','offer_type','price','ab_test',
              'vehicle_type','registration_year','gearbox','power_ps','model',
              'odometer','registration_month','fuel_type','brand','unrepaired_damage'
              ,'ad_created','num_photos','postal_code','last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

So, We made some changes to the columns, After comparing them it's easy to tell that we changed our column names from `CamelCase` to `snake_case`, making our data easier to handle and more clear.

## Initial Exploration and Cleaning
---

In [5]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-21 16:37:21  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

When we use `dataframe.describe()` method, It gives us a lot of information about the numerical columns and non-numerical columns. Upon using the method, and looking at the columns `seller` and `offer_type`, we see there are only 2 unique values and the fequency of the most used value is `49999` in both cases. So, there is no point in keeping these columns anymore, so we will remove them. The column `num_photos` also looks odd, So ,We'll look into that too.

In [6]:
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

It looks like in `num_photos`, 0 has been chosen for all the values so we'll remove this column too.

In [7]:
autos=autos.drop(['seller','offer_type','num_photos'],axis=1)

Let's take a look at the `dataframe.describe()` method again.

## Exploring the Odometer and Price Columns

In [8]:
autos.describe(include='all')

date_crawled         name  price ab_test vehicle_type  \
count                 50000        50000  50000   50000        44905   
unique                48213        38754   2357       2            8   
top     2016-03-21 16:37:21  Ford_Fiesta     $0    test    limousine   
freq                      3           78   1421   25756        12859   
mean                    NaN          NaN    NaN     NaN          NaN   
std                     NaN          NaN    NaN     NaN          NaN   
min                     NaN          NaN    NaN     NaN          NaN   
25%                     NaN          NaN    NaN     NaN          NaN   
50%                     NaN          NaN    NaN     NaN          NaN   
75%                     NaN          NaN    NaN     NaN          NaN   
max                     NaN          NaN    NaN     NaN          NaN   

        registration_year  gearbox      power_ps  model   odometer  \
count        50000.000000    47320  50000.000000  47242      50000   
unique                NaN        2           NaN    245         13   
top                   NaN  manuell           NaN   golf  150,000km   
freq                  NaN    36993           NaN   4024      32424   
mean          2005.073280      NaN    116.355920    NaN        NaN   
std            105.712813      NaN    209.216627    NaN        NaN   
min           1000.000000      NaN      0.000000    NaN        NaN   
25%           1999.000000      NaN     70.000000    NaN        NaN   
50%           2003.000000      NaN    105.000000    NaN        NaN   
75%           2008.000000      NaN    150.000000    NaN        NaN   
max           9999.000000      NaN  17700.000000    NaN        NaN   

        registration_month fuel_type       brand unrepaired_damage  \
count         50000.000000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   
mean              5.723360       NaN         NaN               NaN   
std               3.711984       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN

We can see that `price` and `odometer` columns are object type, even though the values are numeric, that is because of the symbols attached to the values. Let's convert these columns to numeric type so we can perform more operations on them and use them for our analysis, such as `mean()`, `max()`, `min()`. We will use the `series.str.replace` function to remove the symbols and then we wil convert them in `float` or `int`.

In [9]:
autos['price']=(autos['price']      # removing unwanted symbols from column and changing datatypes
                    .str.replace('$','')
                    .str.replace(',','')
                    .astype(int)
               )
autos['odometer']=(autos['odometer']# removing unwanted symbols from column and changing datatypes
                       .str.replace(",",'')
                       .str.replace("km",'')
                       .astype(int)
                  )
autos.rename({'odometer':'odometer_km'},axis=1,inplace=True)
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0       5000  control          bus               2004    manuell       158   
1       8500  control    limousine               1997  automatik       286   
2       8990     test    limousine               2009    manuell       102   
3       4350  control   kleinwagen               2007  automatik        71   
4       1350     test        kombi               2003    manuell         0   
...      ...      ...          ...                ...        ...       ...   
49995  24900  control    limousine               2011  automatik       239   
49996   1980  control       cabrio               1996    manuell        75   
49997  13200     test       cabrio               2014  automatik        69   
49998  22900  control        kombi               2013    manuell       150   
49999   1250  control    limousine               1996    manuell       101   

        model  odometer_km  registration_month fuel_type       brand  \
0      andere       150000                   3       lpg     peugeot   
1         7er       150000                   6    benzin         bmw   
2        golf        70000                   7    benzin  volkswagen   
3      fortwo        70000                   6    benzin       smart   
4       focus       150000                   7    benzin        ford   
...       ...          ...                 ...       ...         ...   
49995      q5       100000                   1    diesel        audi   
49996   astra       150000                   5    benzin        opel   
49997     500         5000                  11    benzin        fiat   
49998      a3        40000                  11    diesel        audi   
49999  vectra       150000                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[50000 rows x 17 columns]

So we converted both columns to `numeric` types, and now we can see the `count`,`mean` and all the information about the them, which helps a great deal in the analysis. We also changed the name of the `odometer` column to `odometer_km`, since we removed the `'km'` from the values.

In [10]:
autos.describe()

price  registration_year      power_ps    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month   postal_code  
count        50000.000000  50000.000000  
mean             5.723360  50813.627300  
std              3.711984  25779.747957  
min              0.000000   1067.000000  
25%              3.000000  30451.000000  
50%              6.000000  49577.000000  
75%              9.000000  71540.000000  
max             12.000000  99998.000000

In ` price` column, we can see that the `std` variation is really high, which seems odd, and the min value is `0`, which technically should not be possible for the price of a car. Let's check out the `price` column a little more

In [11]:
autos['price'].value_counts().head()

0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64

In [12]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [13]:
autos['price'].value_counts().sort_index().head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

If we look at the `value_counts()` for price and looking at the top 20 value counts in both ascending and descending order, We notice that the offers with value `0` are about `1421`, which doesn't look right but ebay bidding options can start at `1`, As for the counts with highest values, the highest value goes to `100` million dollars, That is pretty unrealistic. we can see that until `350000` values were growing steady, after that there was an unrealistic jump in values. We are going to keep the rows with values between `1` to `350000`.

In [14]:
autos=autos[autos['price'].between(1,350000)]
autos.sort_values(by='price',ascending=False)

date_crawled                                               name  \
36818  2016-03-27 18:37:37                                        Porsche_991   
14715  2016-03-30 08:37:24                 Rolls_Royce_Phantom_Drophead_Coupe   
34723  2016-03-23 16:37:29  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
35923  2016-04-03 07:56:23  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
12682  2016-03-28 22:48:01  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
...                    ...                                                ...   
39507  2016-04-02 12:49:46                          Tausche_gegen_anderen_Pkw   
37223  2016-03-25 10:06:23  Ford_Explorer_USA_4.0l_Allrad_Gelaendewagen_Ja...   
30114  2016-03-11 11:59:20                           Audi_80_Typ89.automatilk   
11314  2016-03-19 12:45:16                     Suche_Autos_mit_und_ohne._Tuev   
21342  2016-03-21 12:53:46                                        Opel_Vectra   

        price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
36818  350000  control        coupe               2016    manuell       500   
14715  345000  control       cabrio               2012  automatik       460   
34723  299000     test        coupe               1977    manuell       260   
35923  295000     test       cabrio               2015  automatik       400   
12682  265000  control        coupe               2016  automatik       500   
...       ...      ...          ...                ...        ...       ...   
39507       1     test          NaN               2016    manuell        75   
37223       1     test          suv               1997  automatik       207   
30114       1  control    limousine               1989        NaN         0   
11314       1     test          NaN               2000        NaN         0   
21342       1  control    limousine               1998    manuell        74   

        model  odometer_km  registration_month fuel_type           brand  \
36818     911         5000                   3    benzin         porsche   
14715     NaN        20000                   8    benzin  sonstige_autos   
34723     911       100000                   7    benzin         porsche   
35923     911         5000                   6    benzin         porsche   
12682     911         5000                   3    benzin         porsche   
...       ...          ...                 ...       ...             ...   
39507    golf       150000                   5       NaN      volkswagen   
37223  andere       150000                   9    benzin            ford   
30114      80       150000                   2    benzin            audi   
11314     NaN       150000                   0       NaN  sonstige_autos   
21342  vectra       150000                   3    benzin            opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
36818              nein  2016-03-27 00:00:00        70499  2016-03-27 18:37:37  
14715              nein  2016-03-30 00:00:00        73525  2016-04-07 00:16:26  
34723              nein  2016-03-23 00:00:00        61462  2016-04-06 16:44:50  
35923              nein  2016-04-03 00:00:00        74078  2016-04-03 08:56:20  
12682              nein  2016-03-28 00:00:00        70193  2016-04-05 03:44:51  
...                 ...                  ...          ...                  ...  
39507              nein  2016-04-02 00:00:00         4229  2016-04-06 10:46:56  
37223                ja  2016-03-25 00:00:00        61276  2016-04-03 21:45:18  
30114               NaN  2016-03-11 00:00:00        51103  2016-03-11 12:45:56  
11314               NaN  2016-03-19 00:00:00        49808  2016-03-20 14:18:09  
21342              nein  2016-03-21 00:00:00        52070  2016-03-26 18:44:41  

[48565 rows x 17 columns]

Now, We'll take a look at the `odometer` column and inspect it.

In [15]:
autos['odometer_km'].value_counts()

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
30000       780
20000       762
10000       253
Name: odometer_km, dtype: int64

All the values look rounded in the column, additionally there are more high values than low values, But nothing looks abnormal here so, We are gonna leave it like this.

## Exploring the date columns
------



In [16]:
autos.head(1)

date_crawled                              name  price  ab_test  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD   5000  control   

  vehicle_type  registration_year  gearbox  power_ps   model  odometer_km  \
0          bus               2004  manuell       158  andere       150000   

   registration_month fuel_type    brand unrepaired_damage  \
0                   3       lpg  peugeot              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54

We can see that the columns related to date are as follows:
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website
We know which ones are form the website and crawler by referring to the data dictionary.

Let's check their data types.

In [17]:
autos.dtypes.loc[['date_crawled','last_seen','ad_created','registration_month','registration_year']]

date_crawled          object
last_seen             object
ad_created            object
registration_month     int64
registration_year      int64
dtype: object

In [18]:
autos[['date_crawled','ad_created','last_seen']].head(5)

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

So, Now that we have extracted all three columns,namely `date_crawled`,`ad_created`, and `last_seen`, We can see that their dates are arranged in order of `YYYY-MM-DD H:M:S`, If we want to check the distribution of values according to their dates, We can use `str.[:10]` method with the column that we want. We got the first 10 letters of the string because that is the date. After that we can use the `value_counts()` method and then we can sort the dates using `sort_index()` method. We will now see the precentage of entries for a single date using the `normalize=True` argument of `value_counts()` method that returns the counts in proportions instead of counts.

In [19]:
(autos['date_crawled'].str[:10] #first 10 letter gives us the dates
     .value_counts(normalize=True,dropna=False) #normalize argument gave us proportions
     .sort_index()*100   # sorted them in order of oldest to latest
                        )

2016-03-05    2.532688
2016-03-06    1.404304
2016-03-07    3.601359
2016-03-08    3.329558
2016-03-09    3.308967
2016-03-10    3.218367
2016-03-11    3.257490
2016-03-12    3.691959
2016-03-13    1.566972
2016-03-14    3.654896
2016-03-15    3.428395
2016-03-16    2.960980
2016-03-17    3.162772
2016-03-18    1.291053
2016-03-19    3.477813
2016-03-20    3.788737
2016-03-21    3.737259
2016-03-22    3.298672
2016-03-23    3.222485
2016-03-24    2.934212
2016-03-25    3.160712
2016-03-26    3.220426
2016-03-27    3.109235
2016-03-28    3.486050
2016-03-29    3.409863
2016-03-30    3.368681
2016-03-31    3.183363
2016-04-01    3.368681
2016-04-02    3.547823
2016-04-03    3.860805
2016-04-04    3.648718
2016-04-05    1.309585
2016-04-06    0.317101
2016-04-07    0.140019
Name: date_crawled, dtype: float64

Looks like the crawling period lasted about one month from March 03,2016 to April 07,2016. THe proportions of listings crawled over the period look uniform.

In [20]:
(autos['last_seen'].str[:10]  #first 10 letter gives us the dates
.value_counts(normalize=True,dropna=False) #normalize argument gave us proportions
.sort_index()*100 # sorted them in order of oldest to latest
                        )

2016-03-05     0.107073
2016-03-06     0.432410
2016-03-07     0.539483
2016-03-08     0.741275
2016-03-09     0.959539
2016-03-10     1.066612
2016-03-11     1.237517
2016-03-12     2.378256
2016-03-13     0.889529
2016-03-14     1.260167
2016-03-15     1.587563
2016-03-16     1.645218
2016-03-17     2.808607
2016-03-18     0.735097
2016-03-19     1.583445
2016-03-20     2.065273
2016-03-21     2.063214
2016-03-22     2.137342
2016-03-23     1.853186
2016-03-24     1.976732
2016-03-25     1.921137
2016-03-26     1.680222
2016-03-27     1.564913
2016-03-28     2.085864
2016-03-29     2.234119
2016-03-30     2.477093
2016-03-31     2.378256
2016-04-01     2.279419
2016-04-02     2.491506
2016-04-03     2.520334
2016-04-04     2.448265
2016-04-05    12.476063
2016-04-06    22.180583
2016-04-07    13.194688
Name: last_seen, dtype: float64

In this column, it determines when a listing was last seen by the crawler, which means the listing was removed, most likely because the car was sold.

We see a spike at the end of the crawling period, the spike is not likely an increase in sales because it does not make sense for the sales to grow like that suddenly, We see that the ending dates line up with the end of crawling period, so that makes sense.

In [21]:
(autos['ad_created'].str[:10]  #first 10 letter gives us the dates
    .value_counts(normalize=True,dropna=False) #normalize argument gave us proportions
    .sort_index()*100 # sorted them in order of oldest to latest
                        )

2015-06-11    0.002059
2015-08-10    0.002059
2015-09-09    0.002059
2015-11-10    0.002059
2015-12-05    0.002059
                ...   
2016-04-03    3.885514
2016-04-04    3.685782
2016-04-05    1.181921
2016-04-06    0.325337
2016-04-07    0.125605
Name: ad_created, Length: 76, dtype: float64

the `ad_creation` dates vary a lot, A lot of them were created within a month or two of the last seen date, but some were created a while ago, the oldest one is almost 9 months before the ad was created

Now, Let's take a look at the `registration_year` column.

In [22]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

We see that the maximum value in the registration column is dated 9999 year, which is impossible and the minimum is dated on year 1000, when cars or even internet was not invented, so that doesn't make sense. Next We'll fix the registration year column so we can go ahead with our analysis!

## Dealing with Incorrect Registration Year Data 

In [23]:
autos['registration_year'].sort_values(ascending=False).head(10)

33950    9999
38076    9999
8012     9999
49910    9000
25003    8888
8360     6200
27618    5911
24519    5000
49153    5000
22799    5000
Name: registration_year, dtype: int64

In [24]:
autos['registration_year'].sort_values().head(6)

22316    1000
49283    1001
24511    1111
32585    1800
10556    1800
22659    1910
Name: registration_year, dtype: int64

Looking at the data tables above, We can see many values in the `registration_year` that are really odd, like `1000` and `9999`.

We know that the the `last_seen` date by crawler was `2016`, so any year that comes after `2016` is definitely not valid. Now, the dates before 2016, can be a bit hard to manage. Here, We know that the earliest cars would be in the earlier decades of 1900's atmost. So we will look for rows that have a `registration_year` between 1900-2016,inclusive. We'll decide later if we want to drop the rows with impossible `registration_year` or not later.

In [25]:
(autos[autos['registration_year'].between(1900,2016)]
        .loc[:,'registration_year']
        .value_counts(normalize=True)
        ).sort_index()*100

1910    0.010711
1927    0.002142
1929    0.002142
1931    0.002142
1934    0.004284
          ...   
2012    2.806281
2013    1.720186
2014    1.420278
2015    0.839742
2016    2.613483
Name: registration_year, Length: 78, dtype: float64

This is the pecentage of entries for each individual year between 1900-2016,inclusive.

Now, We have to decide if we want to keep the rows and change their `registration_year` to `np.nan` or we want to remove those rows, because their `registration_year` is definitely inaccurate. Let's see how much data has inaccurate year of registration.

In [26]:
(autos[~autos['registration_year'].between(1900,2016)].size)/ autos.size

0.038793369710697

In [27]:
autos.size

825605

Hmmm, We see that only 3 percent of the data has inaccurate `registration_year`, so it is best that we remove them. Let's modify our data.

In [28]:
autos=autos[autos['registration_year'].between(1900,2016)]

In [29]:
autos.size

793577

In [30]:
autos.head(1)

date_crawled                              name  price  ab_test  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD   5000  control   

  vehicle_type  registration_year  gearbox  power_ps   model  odometer_km  \
0          bus               2004  manuell       158  andere       150000   

   registration_month fuel_type    brand unrepaired_damage  \
0                   3       lpg  peugeot              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54

Let's move onto our next step!

## Exploring Price by Brand

In [31]:
autos[['price','brand','vehicle_type','gearbox']]

price       brand vehicle_type    gearbox
0       5000     peugeot          bus    manuell
1       8500         bmw    limousine  automatik
2       8990  volkswagen    limousine    manuell
3       4350       smart   kleinwagen  automatik
4       1350        ford        kombi    manuell
...      ...         ...          ...        ...
49995  24900        audi    limousine  automatik
49996   1980        opel       cabrio    manuell
49997  13200        fiat       cabrio  automatik
49998  22900        audi        kombi    manuell
49999   1250        opel    limousine    manuell

[46681 rows x 4 columns]

In [32]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

Above, We can see all the unique brands that are in our dataset. We want to know the high-end brands being lsited on the site, and we want to find out what is the mean value of the high-end models.

So, what we are going to do is take the the brands whose total price value is more than the mean of `price` column by at least 50%. Okay lets get Started, we will be doing it like this:
* we are going to make a dictionary which has the `brand` as key and the brand's `mean` price as `value` in our dictionary. 
* we are then going to compare each value of the dictionary with (the overall mean value of `price` column + `50 %`) and select the ones with higher values.

In [33]:
brand_price_dict={} #dictionary representing mean value of each brand

brand_mean_over_50={} # dictionary representing brands with mean values 50% 
                      # higher than average mean.

for brand in autos['brand'].unique(): 
    mean_price=autos[autos['brand']==brand]['price'].mean() #using boolean indexing
    brand_price_dict[brand]=mean_price #first dict created

mean_entire_column=autos['price'].mean() #avg mean of price column

for brand in brand_price_dict:
    if brand_price_dict[brand] > (mean_entire_column+(0.5*mean_entire_column)):
           brand_mean_over_50[brand]=brand_price_dict[brand].round(2) #rounded to
                                                                      #2 decimals
brand_mean_over_50
     

{'audi': 9336.69,
 'sonstige_autos': 12338.55,
 'porsche': 45643.94,
 'mini': 10613.46,
 'jeep': 11650.5,
 'jaguar': 11635.49,
 'land_rover': 19108.09}

The brand with highest mean value is `Porsche`, with almost 2.5 times the mean value of `land_rover` which is the second highest. Let's find out the top 5 listings of `Porsche` and `land_rover`.

In [34]:
(autos[autos['brand']=='porsche']
    .sort_values(by='price',ascending=False)
    .head(5)
    )

date_crawled                                               name  \
36818  2016-03-27 18:37:37                                        Porsche_991   
34723  2016-03-23 16:37:29  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
35923  2016-04-03 07:56:23  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
12682  2016-03-28 22:48:01  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
37840  2016-03-21 10:50:12                                        Porsche_997   

        price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
36818  350000  control        coupe               2016    manuell       500   
34723  299000     test        coupe               1977    manuell       260   
35923  295000     test       cabrio               2015  automatik       400   
12682  265000  control        coupe               2016  automatik       500   
37840  220000     test        coupe               2008    manuell       415   

      model  odometer_km  registration_month fuel_type    brand  \
36818   911         5000                   3    benzin  porsche   
34723   911       100000                   7    benzin  porsche   
35923   911         5000                   6    benzin  porsche   
12682   911         5000                   3    benzin  porsche   
37840   911        30000                   7    benzin  porsche   

      unrepaired_damage           ad_created  postal_code            last_seen  
36818              nein  2016-03-27 00:00:00        70499  2016-03-27 18:37:37  
34723              nein  2016-03-23 00:00:00        61462  2016-04-06 16:44:50  
35923              nein  2016-04-03 00:00:00        74078  2016-04-03 08:56:20  
12682              nein  2016-03-28 00:00:00        70193  2016-04-05 03:44:51  
37840              nein  2016-03-21 00:00:00        69198  2016-04-06 04:46:14

In [35]:
(autos[autos['brand']=='land_rover']
    .sort_values(by='price',ascending=False)
    .head(5)
    )

date_crawled                                               name  \
2184   2016-03-06 04:03:08                                   Land_Rover_Vouge   
28962  2016-03-31 11:54:16          Land_Rover_Range_Rover_SDV8_Autobiography   
23342  2016-03-21 13:37:16      Land_Rover_Range_Rover_Sport_SDV8_HSE_Dynamic   
40026  2016-03-07 13:46:05  Land_Rover_Defender_110_DPF_Station_Wagon_Heri...   
33790  2016-03-05 19:56:22  Land_Rover_Range_Rover_Sport_TDV6_HSE_7_Si._AH...   

       price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
2184   99000     test          suv               2004  automatik       220   
28962  86500  control          suv               2013  automatik       340   
23342  74900     test          suv               2014  automatik       340   
40026  70000  control          suv               2015    manuell       122   
33790  61900     test          suv               2014  automatik       258   

                   model  odometer_km  registration_month fuel_type  \
2184              andere       150000                   6    diesel   
28962        range_rover        50000                   4    diesel   
23342  range_rover_sport        30000                   4    diesel   
40026           defender         5000                  12    diesel   
33790  range_rover_sport        30000                   4    diesel   

            brand unrepaired_damage           ad_created  postal_code  \
2184   land_rover              nein  2016-03-06 00:00:00        13357   
28962  land_rover              nein  2016-03-31 00:00:00        80803   
23342  land_rover              nein  2016-03-21 00:00:00        28217   
40026  land_rover              nein  2016-03-07 00:00:00        44137   
33790  land_rover              nein  2016-03-05 00:00:00        48167   

                 last_seen  
2184   2016-04-07 10:46:41  
28962  2016-04-06 04:46:42  
23342  2016-04-06 08:45:50  
40026  2016-04-06 05:46:25  
33790  2016-03-07 06:45:25

Whew, Those are some nice cars! Okay Let's move on to our next part.

##  Storing Aggregate Data in a DataFrame
Now that we have our list of expensive cars, We want to check their odometer readings and compare them to their mean prices to find out to what extent the odometer readings affect the prices of cars.

To do this, we will repeat the process we did for `mean` prices, to find out the `mean odometer readings` for the high-end brands and create a seperate `dictionary` for that. To compare the mean reading with mean prices, We will create a dataframe that has both the dictionaries side by side. We will be using:
* pandas series constructor
* pandas dataframe constructor
to acheive the task.

In [36]:
brand_odometer_mean={}

for brand in brand_mean_over_50: #dictionary of mean prices of high-end brands
    odometer_mean= autos[autos['brand']==brand]['odometer_km'].mean().round(2)
    brand_odometer_mean[brand]=odometer_mean
brand_odometer_mean


{'audi': 129157.39,
 'sonstige_autos': 89956.33,
 'porsche': 96853.15,
 'mini': 88105.13,
 'jeep': 127122.64,
 'jaguar': 124178.08,
 'land_rover': 118010.2}

Now using the two dictionaries, We will create two seperate pandas dataframes and concatenate them. First we will convert the dictioanries in seperate series and then we will be able to convert them into dataframes.

In [37]:
# converting it into series
brand_odometer_mean=pd.Series(brand_odometer_mean) 

#converting it into dataframe
brand_odometer_mean=pd.DataFrame(brand_odometer_mean,columns=['mean_odometer_km'])
    
brand_odometer_mean

mean_odometer_km
audi                   129157.39
sonstige_autos          89956.33
porsche                 96853.15
mini                    88105.13
jeep                   127122.64
jaguar                 124178.08
land_rover             118010.20

We will do the same with `brand_mean_over_50`

In [38]:
# converting it into series
brand_mean_over_50=pd.Series(brand_mean_over_50)

#converting it into dataframe
brand_mean_over_50=pd.DataFrame(brand_mean_over_50,columns=['mean_price'])

brand_mean_over_50


mean_price
audi               9336.69
sonstige_autos    12338.55
porsche           45643.94
mini              10613.46
jeep              11650.50
jaguar            11635.49
land_rover        19108.09

Now we will `concatenate` them to create a single dataframe. In this case our list is small so we are able to compare them visually but when the dictionary is really big, it will be really useful to convert them into dataframes so we can even perform many operations and find maximums and minimums easily.

In [39]:
price_vs_odometer=pd.concat([brand_mean_over_50,brand_odometer_mean],axis=1)
price_vs_odometer

mean_price  mean_odometer_km
audi               9336.69         129157.39
sonstige_autos    12338.55          89956.33
porsche           45643.94          96853.15
mini              10613.46          88105.13
jeep              11650.50         127122.64
jaguar            11635.49         124178.08
land_rover        19108.09         118010.20

Let's make it pretty and get rid of the decimal places.

In [40]:
price_vs_odometer=price_vs_odometer.astype(int)
price_vs_odometer

mean_price  mean_odometer_km
audi                  9336            129157
sonstige_autos       12338             89956
porsche              45643             96853
mini                 10613             88105
jeep                 11650            127122
jaguar               11635            124178
land_rover           19108            118010

Based on the data we have, We can analyze the price dynamic of mileage with price in the used car market. Here it seems like the cheaper cars have on-avergae high mileage which means they have been used more than the expensive ones. Since, We were considering the high-end models only, it creates a bias in price towards the really expensive cars like `porsche`, However, when brands with some-what similar mean price are chosen, it will give a great analysis, such as , Which brand car owners were able to use their cars the most before they felt the need to change their car or sell it, or Which brand prices drop most with mileage and which ones stay the same.

***AG***